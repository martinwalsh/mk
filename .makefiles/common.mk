###############################################################################
# COMMON HELPERS
###############################################################################
.DEFAULT_GOAL := help
MAKEFLAGS += --no-print-directory

###############################################################################
# LOGGING HELPERS
#

ifneq (,$(shell tput colors 2> /dev/null))
tput   := $(shell command -v tput 2>/dev/null)
bold   := $(shell $(tput) bold)
reset   := $(shell $(tput) sgr0)

red    := $(shell $(tput) setaf 1)
green  := $(shell $(tput) setaf 2)
yellow := $(shell $(tput) setaf 3)
cyan   := $(shell $(tput) setaf 6)
gray   := $(shell $(tput) setaf 248)
endif

# $(call dbg,<message>)
#
# Returns <message> wrapped in tput bold formatting, in the color gray.
define dbg
echo "$(bold)$(gray)""$(strip $(1))""$(reset)"
endef

# $(call log,<message>)
#
# Returns <message> wrapped in tput bold formatting.
define log
echo "$(bold)""$(strip $(1))""$(reset)"
endef

# $(call wrn,<message>)
#
# Returns <message> wrapped in tput bold formatting, in the color yellow.
define wrn
echo "$(bold)$(yellow)""$(strip $(1))""$(reset)"
endef

# $(call err,<message>)
#
# Returns <message> wrapped in tput bold formatting, in the color red.
define err
echo "$(bold)$(red)""$(strip $(1))""$(reset)"
endef

# $(call raise,<message>)
#
# Raises an error with the given <message> wrapped in tput bold formatting, in the color red.
define raise
$(error $(bold)$(red)$(strip $(1))$(reset))
endef


###############################################################################
# STRING HELPERS
#

# $(call lc,ALLUPPERCASE) -> alluppercase
lc = $(shell echo "$(1)" | tr '[:upper:]' '[:lower:]')

# $(call uc,alluppercase) -> ALLUPPERCASE
uc = $(shell echo "$(1)" | tr '[:lower:]' '[:upper:]'

###############################################################################
# PREREQUISITE HELPERS
#

# some_target: | _env_HOME
# 	@echo "do something"
#
# Prerequisite target that errors if the given environment variable is not
# defined in the current context, as named by the wildcard part of the target name.
_env_%:
	@_='$(or $($*),$(error `$*` is required))'


# joke: | _cmd_curl
# 	@curl -sSfL https://icanhazdadjoke.com/
# .PHONY: joke
#
# Prerequisite target that errors if the given program is not
# found on the system PATH, as named in the wildcard part of the target name.
_cmd_%:
	@_='$(or $(shell command -v $* 2>/dev/null),$(error `$*` command not found))'


###############################################################################
# PATH HELPERS
#

# $(call assert-path, some/path/to/a/folder/or/file, ERROR: Path does not exist)
#
# Callable that checks if a path exists, or performs an action (e.g. raise an error).
define assert-path
@_='$(or $(wildcard $(strip $(1))),$(call raise,$(or $(2),Path $(1) does not exist)))'
endef


###############################################################################
# OS/PLATFORM DETECTION
#

ifeq ($(OS),Windows_NT)
OS_NAME := windows
else
OS_NAME := $(call lc,$(shell uname -s))
endif

###############################################################################
# MISC HELPERS
#

# replace-text: | _cmd_sed
# 	$(call sed_i) 's/old/new/g' file.txt
#
# Callable that returns an appropriate `sed` in-place command for the current platform.
_gnu_sed := $(shell sed --version 2>/dev/null | grep -o 'GNU' | head -1)
sed_i := sed $(if $(filter $(call _gnu_sed),GNU),-i'',-i '')


# $(call with_dryrun, docker-compose push)
#
# Callable that prevents the execution of the given
# command when DR is set, by logging the command instead.
define with_dryrun
$(if $(DR),$(call wrn,[DRYRUN] WOULD RUN $(1)),$(1))
endef

# lint: | _fixme
# 	@echo "do something"
#
# Prerequisite target that errors if a FIXME comment is found anywhere in the source.
_fixme: _cmd_git
	@if git grep FIXME -- ':!\.makefiles/**'; then \
		$(call err,One or more FIXME comments are present in the source code.); \
		exit 1; \
	fi
.PHONY: _fixme

# lint: | _todo
# 	@echo "do something"
#
# Prerequisite target that errors if a TODO comment is found anywhere in the source.
_todo: _cmd_git
	@if git grep TODO -- ':!\.makefiles/**'; then \
		$(call err,One or more TODO comments are present in the source code.); \
		exit 1; \
	fi
.PHONY: _todo

###############################################################################
# HELP
#

# #| Print this helpful message
# help::
# 	@$(if $(shell command -v mk 2>/dev/null), mk --help)
# .PHONY: help


###############################################################################
# CLEAN
#

#| Remove all build artifacts
clean::
	@git clean -ffdx
.PHONY: clean
