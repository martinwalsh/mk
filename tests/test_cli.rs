use assert_cmd::prelude::*;
use insta_cmd::assert_cmd_snapshot;
use predicates::prelude::*;
use std::error::Error;

use crate::common::{dedent, Space};
mod common;

#[test]
fn returns_an_error_if_the_makefile_is_missing() -> Result<(), Box<dyn Error>> {
    let space = Space::new();

    space
        .bin()
        .assert()
        .failure()
        .stderr(predicate::str::contains("No Makefile found"));

    Ok(())
}

#[test]
fn traversal_stops_at_the_dot_git_directory() -> Result<(), Box<dyn Error>> {
    let space = Space::new();

    space
        .bin()
        .env("RUST_LOG", "DEBUG")
        .assert()
        .failure()
        .stderr(predicate::str::contains(
            "ERROR mk] No Makefile found in the project!",
        ))
        .stderr(predicate::str::contains(
            "DEBUG mk::makefile] Reached the root of the project.",
        ));

    Ok(())
}

#[test]
fn prints_minimal_help_if_makefile_is_empty_and_no_args_given() -> Result<(), Box<dyn Error>> {
    let space = Space::from_fixture("blank")?;
    assert_cmd_snapshot!(space.bin());
    Ok(())
}

#[test]
fn prints_minimal_help_if_makefile_is_empty_and_h_flag_is_given() -> Result<(), Box<dyn Error>> {
    let space = Space::from_fixture("blank")?;
    assert_cmd_snapshot!(space.bin().arg("-h"));
    Ok(())
}

#[test]
fn prints_minimal_help_if_makefile_is_empty_and_help_flag_is_given() -> Result<(), Box<dyn Error>> {
    let space = Space::from_fixture("blank")?;
    assert_cmd_snapshot!(space.bin().arg("--help"));
    Ok(())
}

#[test]
fn can_execute_make_with_valid_syntax_repeated_targets() -> Result<(), Box<dyn Error>> {
    let space = Space::from_fixture("repeats")?;

    assert_cmd_snapshot!(space.bin().arg("test"));

    space
        .bin()
        .arg("test")
        .assert()
        .success()
        .stdout(predicate::str::contains("This is a test var: success"));

    Ok(())
}

#[test]
fn can_take_make_flags_like_force_build() -> Result<(), Box<dyn Error>> {
    let space = Space::from_fixture("test")?;

    space
        .bin()
        .arg("test")
        .assert()
        .success()
        .stdout(predicate::str::contains("This is a successful test."));

    space
        .bin()
        .arg("test")
        .assert()
        .success()
        .stdout(predicate::str::contains(" is up to date."));

    space
        .bin()
        .arg("test")
        .arg("-B")
        .assert()
        .success()
        .stdout(predicate::str::contains("This is a successful test."));

    Ok(())
}

#[test]
fn uses_env_vars_preceding_the_command() -> Result<(), Box<dyn Error>> {
    let space = Space::from_fixture("envvars")?;

    assert_cmd_snapshot!(space.bin().env("ENVVAR", "success").arg("test"));

    Ok(())
}

#[test]
fn uses_env_vars_passed_as_arguments() -> Result<(), Box<dyn Error>> {
    let space = Space::from_fixture("envvars")?;

    assert_cmd_snapshot!(space.bin().arg("test").arg("ENVVAR=success"));

    Ok(())
}

#[test]
fn shows_make_error_if_an_invalid_target_is_used() -> Result<(), Box<dyn Error>> {
    let space = Space::from_fixture("test")?;

    space.bin().arg("invalid").assert().failure().stderr(
        predicate::str::is_match(r###"\*\*\* No rule to make target [`']invalid'"###).unwrap(),
    ); // Some versions of make use a leading backtick, some don't.

    Ok(())
}

#[test]
fn shows_help_with_documentation() -> Result<(), Box<dyn Error>> {
    let space = Space::from_fixture("docs")?;
    assert_cmd_snapshot!(space.bin().arg("--help"));

    space
        .bin()
        .arg("--help")
        .assert()
        .success()
        .stdout(predicate::str::contains("test  Runs the tests"));

    Ok(())
}

#[test]
fn skips_help_for_targets_without_documentation() -> Result<(), Box<dyn Error>> {
    let space = Space::from_fixture("nodocs")?;

    assert_cmd_snapshot!(space.bin().arg("--help"));

    space
        .bin()
        .arg("--help")
        .assert()
        .success()
        .stdout(predicate::str::contains("test-me").count(0));

    Ok(())
}

#[test]
fn can_parse_help_from_includes() -> Result<(), Box<dyn Error>> {
    let space = Space::from_fixture("include")?;

    let content = dedent(
        r###"
        #| Included test target
        included:
        \techo 'This is another successful test.'
        .PHONY: included
    "###,
    )
    .replace(r"\t", "\t");
    space.add_include("includes", "include", &content)?;

    assert_cmd_snapshot!(space.bin().arg("--help"));

    Ok(())
}

#[test]
fn can_run_included_commands_from_a_subfolder() -> Result<(), Box<dyn Error>> {
    let space = Space::from_fixture("include")?;

    let content = dedent(
        r###"
        #| Included test target
        included:
        \t@echo 'This is another successful test.'
        .PHONY: included
    "###,
    )
    .replace(r"\t", "\t");
    space.add_include("includes", "include", &content)?;

    assert_cmd_snapshot!(space.bin().arg("included"));

    Ok(())
}
