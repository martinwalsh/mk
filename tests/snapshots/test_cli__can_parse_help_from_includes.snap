---
source: tests/test_cli.rs
info:
  program: mk
  args:
    - "--help"
---
success: true
exit_code: 0
----- stdout -----
An experimental command-line wrapper for GNU Make.

Usage: mk [OPTIONS] [COMMAND]

Commands:
  test      Parent test target
  included  Included test target

Options:
      --me       Run an internal `mk` command
  -h, --help     Print help
  -V, --version  Print version

----- stderr -----
